File: src/providers/claude.zig
const std = @import("std");
const HttpClient = @import("../api/http_client.zig").HttpClient;
const HttpResponse = @import("../api/http_client.zig").HttpResponse;

pub const ClaudeProvider = struct {
    http_client: HttpClient,
    api_key: []const u8,
    base_url: []const u8,
    model: []const u8,
    allocator: std.mem.Allocator,
    
    pub fn init(allocator: std.mem.Allocator, api_key: []const u8) ClaudeProvider {
        return .{
            .http_client = HttpClient.init(allocator),
            .api_key = api_key,
            .base_url = "https://api.anthropic.com/v1",
            .model = "claude-3-5-sonnet-20241022",
            .allocator = allocator,
        };
    }
    
    pub fn deinit(self: *ClaudeProvider) void {
        self.http_client.deinit();
    }
    
    pub fn chatCompletion(self: *ClaudeProvider, messages: []const ChatMessage, conversation_id: []const u8) !ChatCompletionResponse {
        _ = conversation_id;
        
        // Convert messages to Claude format (system message separate)
        var system_message: ?[]const u8 = null;
        var user_messages = std.ArrayList(ClaudeMessage).init(self.allocator);
        defer user_messages.deinit();
        
        for (messages) |message| {
            if (std.mem.eql(u8, message.role, "system")) {
                system_message = message.content;
            } else {
                try user_messages.append(.{
                    .role = message.role,
                    .content = message.content,
                });
            }
        }
        
        // Create request payload
        var payload = std.ArrayList(u8).init(self.allocator);
        defer payload.deinit();
        
        const writer = payload.writer();
        try writer.print("{{\"model\":\"{s}\",\"max_tokens\":4000,\"messages\":[", .{self.model});
        
        for (user_messages.items, 0..) |message, i| {
            if (i > 0) try writer.print(",");
            try writer.print("{{\"role\":\"{s}\",\"content\":", .{message.role});
            try std.json.stringify(message.content, .{}, writer);
            try writer.print("}}");
        }
        
        try writer.print("]");
        
        if (system_message) |sys_msg| {
            try writer.print(",\"system\":");
            try std.json.stringify(sys_msg, .{}, writer);
        }
        
        try writer.print("}}");
        
        // Create headers
        var headers = std.StringHashMap([]const u8).init(self.allocator);
        defer headers.deinit();
        try headers.put("x-api-key", self.api_key);
        try headers.put("Content-Type", "application/json");
        try headers.put("anthropic-version", "2023-06-01");
        
        // Make HTTP request
        const url = try std.fmt.allocPrint(self.allocator, "{s}/messages", .{self.base_url});
        defer self.allocator.free(url);
        
        var response = self.http_client.post(url, payload.items, headers) catch |err| {
            std.log.err("Claude HTTP request failed: {}", .{err});
            return ChatCompletionError.NetworkError;
        };
        defer response.deinit();
        
        if (!response.isSuccess()) {
            std.log.err("Claude API error: status {}, body: {s}", .{ response.status, response.body });
            return switch (response.status) {
                401 => ChatCompletionError.AuthenticationFailed,
                429 => ChatCompletionError.RateLimited,
                400...499 => ChatCompletionError.BadRequest,
                500...599 => ChatCompletionError.ServerError,
                else => ChatCompletionError.NetworkError,
            };
        }
        
        // Parse response
        const parsed = std.json.parseFromSlice(ClaudeResponse, self.allocator, response.body, .{}) catch |err| {
            std.log.err("Failed to parse Claude response: {}, body: {s}", .{ err, response.body });
            return ChatCompletionError.InvalidResponse;
        };
        defer parsed.deinit();
        
        if (parsed.value.content.len == 0) {
            return ChatCompletionError.InvalidResponse;
        }
        
        // Find text content
        var text_content: ?[]const u8 = null;
        for (parsed.value.content) |content| {
            if (std.mem.eql(u8, content.type, "text")) {
                text_content = content.text;
                break;
            }
        }
        
        if (text_content == null) {
            return ChatCompletionError.InvalidResponse;
        }
        
        const content = try self.allocator.dupe(u8, text_content.?);
        
        return ChatCompletionResponse{
            .content = content,
            .model = try self.allocator.dupe(u8, parsed.value.model),
            .usage = Usage{
                .prompt_tokens = parsed.value.usage.input_tokens,
                .completion_tokens = parsed.value.usage.output_tokens,
                .total_tokens = parsed.value.usage.input_tokens + parsed.value.usage.output_tokens,
            },
        };
    }
    
    pub fn codeCompletion(self: *ClaudeProvider, prompt: []const u8, language: ?[]const u8, context: ?[]const u8) !CodeCompletionResponse {
        var messages = std.ArrayList(ChatMessage).init(self.allocator);
        defer messages.deinit();
        
        // System message for code completion
        const system_prompt = if (language) |lang|
            try std.fmt.allocPrint(self.allocator, 
                "You are an expert {s} programmer. Complete the following code. Return only the completion without explanation or markdown formatting.", 
                .{lang})
        else
            "You are an expert programmer. Complete the following code. Return only the completion without explanation or markdown formatting.";
        defer if (language != null) self.allocator.free(system_prompt);
        
        try messages.append(.{ .role = "system", .content = system_prompt });
        
        // Add context if provided
        if (context) |ctx| {
            const context_msg = try std.fmt.allocPrint(self.allocator, "Context:\n{s}\n\nComplete this code:", .{ctx});
            defer self.allocator.free(context_msg);
            try messages.append(.{ .role = "user", .content = context_msg });
        }
        
        try messages.append(.{ .role = "user", .content = prompt });
        
        const chat_response = try self.chatCompletion(messages.items, "code-completion");
        
        return CodeCompletionResponse{
            .completion = chat_response.content,
            .language = if (language) |lang| try self.allocator.dupe(u8, lang) else null,
            .confidence = 0.9, // Claude typically has very high confidence
        };
    }
    
    pub fn streamChatCompletion(self: *ClaudeProvider, messages: []const ChatMessage, callback: StreamCallback) !void {
        // For now, simulate streaming by calling the regular completion and invoking callback
        const response = try self.chatCompletion(messages, "stream");
        defer self.allocator.free(response.content);
        defer self.allocator.free(response.model);
        
        // Split content into chunks and stream
        const chunk_size = 15; // Slightly larger chunks for Claude
        var pos: usize = 0;
        
        while (pos < response.content.len) {
            const end = @min(pos + chunk_size, response.content.len);
            const chunk = response.content[pos..end];
            
            const stream_chunk = StreamChunk{
                .content = chunk,
                .done = end >= response.content.len,
                .model = response.model,
            };
            
            callback(stream_chunk);
            pos = end;
            
            // Small delay to simulate streaming
            std.time.sleep(30 * std.time.ns_per_ms);
        }
    }
};

pub const ChatMessage = struct {
    role: []const u8, // "system", "user", "assistant"
    content: []const u8,
};

pub const ChatCompletionResponse = struct {
    content: []const u8,
    model: []const u8,
    usage: Usage,
};

pub const CodeCompletionResponse = struct {
    completion: []const u8,
    language: ?[]const u8,
    confidence: f32,
};

pub const Usage = struct {
    prompt_tokens: u32,
    completion_tokens: u32,
    total_tokens: u32,
};

pub const StreamChunk = struct {
    content: []const u8,
    done: bool,
    model: []const u8,
};

pub const StreamCallback = *const fn (chunk: StreamChunk) void;

pub const ChatCompletionError = error{
    NetworkError,
    InvalidResponse,
    AuthenticationFailed,
    RateLimited,
    ServerError,
    BadRequest,
} || std.mem.Allocator.Error;

// Internal Claude API response structures
const ClaudeMessage = struct {
    role: []const u8,
    content: []const u8,
};

const ClaudeResponse = struct {
    id: []const u8,
    type: []const u8,
    role: []const u8,
    model: []const u8,
    content: []ContentBlock,
    stop_reason: []const u8,
    stop_sequence: ?[]const u8,
    usage: ClaudeUsage,
    
    const ContentBlock = struct {
        type: []const u8, // "text"
        text: []const u8,
    };
    
    const ClaudeUsage = struct {
        input_tokens: u32,
        output_tokens: u32,
    };
};

File: src/providers/claude.rs
use async_trait::async_trait;
use reqwest::Client;
use serde_json::json;
use std::env;
use tracing::{debug, error};

use crate::error::{ZekeError, ZekeResult};
use super::{ProviderClient, ChatRequest, ChatResponse, Provider, Usage, ChatMessage};

pub struct ClaudeClient {
    client: Client,
    api_key: String,
    base_url: String,
}

impl ClaudeClient {
    pub fn new() -> ZekeResult<Self> {
        let api_key = env::var("CLAUDE_API_KEY")
            .or_else(|_| env::var("ANTHROPIC_API_KEY"))
            .map_err(|_| ZekeError::provider("CLAUDE_API_KEY or ANTHROPIC_API_KEY environment variable not set"))?;

        Ok(Self {
            client: Client::new(),
            api_key,
            base_url: "https://api.anthropic.com".to_string(),
        })
    }

    pub fn with_api_key(api_key: String) -> Self {
        Self {
            client: Client::new(),
            api_key,
            base_url: "https://api.anthropic.com".to_string(),
        }
    }

    // Convert OpenAI-style messages to Claude format
    fn convert_messages(&self, messages: &[ChatMessage]) -> (String, Vec<serde_json::Value>) {
        let mut system_prompt = String::new();
        let mut claude_messages = Vec::new();

        for message in messages {
            match message.role.as_str() {
                "system" => {
                    if !system_prompt.is_empty() {
                        system_prompt.push('\n');
                    }
                    system_prompt.push_str(&message.content);
                }
                "user" | "assistant" => {
                    claude_messages.push(json!({
                        "role": message.role,
                        "content": message.content
                    }));
                }
                _ => {
                    // Convert other roles to user
                    claude_messages.push(json!({
                        "role": "user",
                        "content": message.content
                    }));
                }
            }
        }

        (system_prompt, claude_messages)
    }
}

#[async_trait]
impl ProviderClient for ClaudeClient {
    async fn chat_completion(&self, request: &ChatRequest) -> ZekeResult<ChatResponse> {
        let url = format!("{}/v1/messages", self.base_url);

        let (system_prompt, messages) = self.convert_messages(&request.messages);

        let mut payload = json!({
            "model": request.model.as_deref().unwrap_or(self.get_default_model()),
            "messages": messages,
            "max_tokens": request.max_tokens.unwrap_or(2048),
            "temperature": request.temperature.unwrap_or(0.7),
        });

        if !system_prompt.is_empty() {
            payload["system"] = json!(system_prompt);
        }

        debug!("Sending request to Claude: {}", url);

        let response = self.client
            .post(&url)
            .header("x-api-key", &self.api_key)
            .header("Content-Type", "application/json")
            .header("anthropic-version", "2023-06-01")
            .json(&payload)
            .send()
            .await?;

        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_else(|_| "Unknown error".to_string());
            error!("Claude API error: {}", error_text);
            return Err(ZekeError::provider(format!("Claude API error: {}", error_text)));
        }

        let response_json: serde_json::Value = response.json().await?;

        let content = response_json["content"][0]["text"]
            .as_str()
            .unwrap_or("")
            .to_string();

        let model = response_json["model"]
            .as_str()
            .unwrap_or(self.get_default_model())
            .to_string();

        let usage = if let Some(usage_data) = response_json.get("usage") {
            Some(Usage {
                prompt_tokens: usage_data["input_tokens"].as_u64().unwrap_or(0) as u32,
                completion_tokens: usage_data["output_tokens"].as_u64().unwrap_or(0) as u32,
                total_tokens: (usage_data["input_tokens"].as_u64().unwrap_or(0) +
                              usage_data["output_tokens"].as_u64().unwrap_or(0)) as u32,
            })
        } else {
            None
        };

        Ok(ChatResponse {
            content,
            model,
            provider: Provider::Claude,
            usage,
        })
    }

    async fn health_check(&self) -> ZekeResult<bool> {
        // Claude doesn't have a simple health endpoint, so we'll make a minimal request
        let url = format!("{}/v1/messages", self.base_url);

        let payload = json!({
            "model": self.get_default_model(),
            "messages": [{"role": "user", "content": "ping"}],
            "max_tokens": 1
        });

        match self.client
            .post(&url)
            .header("x-api-key", &self.api_key)
            .header("Content-Type", "application/json")
            .header("anthropic-version", "2023-06-01")
            .json(&payload)
            .send()
            .await
        {
            Ok(response) => Ok(response.status().is_success() || response.status().as_u16() == 400),
            Err(_) => Ok(false),
        }
    }

    fn get_provider(&self) -> Provider {
        Provider::Claude
    }

    fn get_default_model(&self) -> &str {
        "claude-3-5-sonnet-20241022"
    }
}